src/routing/mod.rs

```diff
@@ -312,17 +311,42 @@ impl Routing {
         dst: DstLocation,
         content: Bytes,
     ) -> Result<()> {
+        if let DstLocation::EndUser(EndUser::Client {
+            socket_id,
+            public_key,
+        }) = dst
+        {
+            let mut socket = None;
+            {
+                if let Some(socket_addr) = self.stage.state.lock().await.get_socket_addr(&socket_id)
+                {
+                    socket = Some(socket_addr.clone());
+                }
+            }
```

Comment:
```suggestion
            let socket_addr = self.stage.state.lock().await.get_socket_addr(&socket_id).clone();
```

Reply:
```let socket_addr = self.stage.state.lock().await.get_socket_addr(&socket_id).clone();
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                    - temporary value is freed at the end of this statement
    |                               |
    |                               creates a temporary which is freed while still in use
320 | 
321 |             if let Some(socket_addr) = socket_addr {
    |                                        ----------- borrow later used here```

Reply:
That was the case with the if condition. Once this is assigned to socket_addr (after .clone() or copy `*`)
The mutex lock will be released.

```
if let Some(...) = self.stage.state.lock().do_something().await {
     self.do_something_else().await()
}
```
Above, the lock is held in the if condition so within the block ( `self.do_something_else()` in this case) if we try to get a hold of the mutex, then it will hang since there is a deadlock.

Reply:
I think we can dereference it and it will no longer be borrowed then? :thinking:

Reply:
Just tried this locally, and this can be done as:
`let socket_addr = self.stage.state.lock().await.get_socket_addr(&socket_id).copied();`

Reply:
`get_socket_addr` returns `Option<&SocketAddr>`. Calling `clone` on it just clones the whole `Option`, not the thing in it. So it still returns `Option<&SocketAddr>` and therefore is still borrowing from a temporary. @lionel1704 's suggestion is the correct one.